core/node/node_framework/src/implementations/layers/vm_runner/protective_reads.rs

```diff
@@ -0,0 +1,71 @@
+use zksync_config::configs::vm_runner::ProtectiveReadsWriterConfig;
+use zksync_types::L2ChainId;
+use zksync_vm_runner::ProtectiveReadsWriter;
+
+use crate::{
+    implementations::resources::pools::{MasterPool, PoolResource},
+    service::{ServiceContext, StopReceiver},
+    task::{Task, TaskId},
+    wiring_layer::{WiringError, WiringLayer},
+};
+
+#[derive(Debug)]
+pub struct ProtectiveReadsWriterLayer {
+    protective_reads_writer_config: ProtectiveReadsWriterConfig,
+    zksync_network_id: L2ChainId,
+}
+
+impl ProtectiveReadsWriterLayer {
+    pub fn new(
+        protective_reads_writer_config: ProtectiveReadsWriterConfig,
+        zksync_network_id: L2ChainId,
+    ) -> Self {
+        Self {
+            protective_reads_writer_config,
+            zksync_network_id,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl WiringLayer for ProtectiveReadsWriterLayer {
+    fn layer_name(&self) -> &'static str {
+        "vm_runner_protective_reads"
+    }
+
+    async fn wire(self: Box<Self>, mut context: ServiceContext<'_>) -> Result<(), WiringError> {
+        let master_pool = context.get_resource::<PoolResource<MasterPool>>().await?;
+
+        let (protective_reads_writer, tasks) = ProtectiveReadsWriter::new(
+            master_pool.get_custom(3).await?,
```

Comment:
please add comment on why 3 connections

Reply:
Actually, thinking about it one more time, it seems like we need `window_size + 2` size pool here. Please verify my logic I left in the comment